### DISCUSSION TITLE
How to pick & prioritize optimization parameters since we can't simply do all of them at once.

### DISCUSSION THREAD

**Discussion by enjoysmath** [EA31337/EA31337#295] [Q&A] [answered]
I've been experimenting for 3days with the optimization parameters in strategy tester of MT5 on the EA37337.  I've tried many different orderings and number of parameters each iteration through the process.  We can't simply select all because combinatorially not even a super computer could get through the optimization step before the universe implodes (or something like that).  

So how do you recommed we perform optimization / backtesting having designed the tool?

**Comment by kenorb** [comment, marked as answer]
## v2.005

I usually optimize one set of params per run, not all of them at once. Also I think fast genetic algoritm could skip some of the good results.

Since all version are slightly different, the following personal recommendations applies to v2.005.

### General

General optimization hints:

- Results: Don't pick strategies which generated lesser trades (like less than 100 per year) despite of the higher profit factor. I would rather choose strategy with 1000+ trades with lesser profit factor for its statistical reliability and proven consistency, than the one with 100 trades with higher profit factor (in this case try to increase the test period for the given strategy). The healthy strategy should generate at least 1 trade per day on average (after filtering), otherwise revise the filter configuration. 
- Read more at: [Optimization hints](https://github.com/EA31337/EA31337/wiki/Optimization-hints).

### Lite

Optimization from scratch:

- Optimize strategies per one timeframe at a time, and disable all other timeframes, e.g. optimize Strategy_H1, then H4, then M30, and so on. In this case you've 30+ interaction per run, in MT5 with 32+ CPU cores, it'll run all tests at once. Recommended period: 2020-2021.
- Combine all selected strategies and re-run. Avoid enabling strategies lesser than M15.
- Tweak results by optimizing strategies per one timeframe and how they work together (M15 and beyond).
- Consider adding M5, then M1 strategy (not recommended at the moment).

To summarize, the general order of optimization is to first optimize strategies separately per timeframe. Then retest with all selected strategies enabled at once. Next optimize filters if available. Compare final results with different modelling and re-run the final setup (broker or platform) with longer periods.

Using Lite version (in comparison to Advanced) can be useful to have a general idea about the strategies (which are good or bad for the given market) without overcomplicating configuration with filters.

### Advanced

Optimization from scratch:

1. Reset filters to optimal defaults (to avoid unnecessary influence by external factors):

    - SOFM (Signal **Open** Filter Method) depends on your overall desired strategy (as this filter will influence all strategies and their initial trades). Reset to 32 (which is the average best) for grid-like trading, set 0 for non-grid trading approach.
    - SCF (Signal **Close** Filter) to 0 (disable).
    - SOTF (Signal Open **Time** Filter) to average best (e.g. 12).
    - SOSF (Signal Open **Strategy** Filter) to 0
    - TFM (**Tick** Filter Method) it depends, but never set it to 0 (it's a mess to not try to filter ticks). If you're optimizing on 1 min OHLCs, set to 32 (open prices), when optimizing by using _Every tick_ modelling, you can pick 7 (ticks at peak). When you change tick filtering after, you may have to redo all the strategies again.
    - Tasks: Disable all (it influences strategy behaviour, so for testing strategy trading logic, tasks are not recommended at initial stage of optimization)
    - OCL/OCP (Order Close **Loss**/**Profit**): set to 0, but it's up to you. Handicapping loss and profit can help to find strategy with more consistency by not relying on some lucky big rewards (or losses) based on spikes at some random past events. It may generate less profit overall (based on past), but you can find it more reliable across the time.
    - OCT (Order Close **Time**) recommended to -30 or -40 (number of bars to close). This avoids orders being opened for too long, because the profit from order based on M15 signal running for few days or weeks is merely based a luck (which is not what we want), not technical signals. So we need to enforce some consistency.

2. Sets all Strategies and Stops to (None). And start optimizing strategies per timeframe.
3. First optimize hourly strategies, then minuted-based one timeframe by one (disabling any other timeframes for quicker run and to avoid unwanted noise), e.g. starting on H1, then H4, then M30, M15 and so on. This allow you to see how they given strategy will perform individually on their own (recommended period: 2020-2021).
4. Then optimize their stops for corresponding timeframes (one timeframe by one): H1 stops for H1 strategy, H4 stops for H4 strategy.
5. To optimize strategies on minute-based timeframes, you have two choices: SOSF (Strategy filter) is set to 0 (minute-based strategies are trading based on their own signals), or to 2 (hourly signals influence minute-based signals, so less trades are opened).
6. When you decide on SOSF=0, you can further optimize using strategies per timeframe (and disabling other), because each strategy's timeframe is not influenced by the other (they trade independently). 
7. When you decide on SOSF=2, then minute-based strategies are influenced by hourly-based strategies (are opened when hourly signals are also confirmed), so keep optimizing minute-based per timeframe at a time, but having H1 and H4 strategies with their stops enabled all the time. In otherwords, you can optimize M30 (where M1-M15 are disabled, but H1 and H4 enabled).
8. When SOSF=2, you can safely optimize strategies lesser than M15, for SOSF=0 optimizing M1-M5 are not recommended (it's going to create a random noise and influencing higher timeframe strategies by using account's margin).
9. Once you selected your active strategies and their stops, re-test all strategies at once how they "play" together. It's important, as one strategies which generates too many trades can obscure and influence other strategies with lesser trades (by using safe margin) plus other factors.
10. Compare results with different modelling: Every tick vs 1 minute OHLC vs Real ticks. If tick modelling is worser than OHLC, try optimizing Tick filter method (currently 7 is recommended). If OHLC results are worser (which is fine), then we don't care as real trading is based on ticks anyway. But remember, playing with tick filter is like playing with a fire (it's unpredictable). If you need simple consistency, you can just use open ticks for now (=32). When optimizing in MT5 for MT4 run, compare the results between platforms as well, and find the average best configuration. Keep in mind that optimization tests of tick filter can be very slow.
11. You can try to polish strategies by running all at once and optimizing one strategy timeframe or their stops per test (selecting multiple enum or filter params is a waste of resources). To save time, you can do that after filters are optimized.
12. Next, optimize strategy filters (one param and its range per test). Recommended order: SOTF, SOFM, SCF. You can repeat some of the params again, after changing the other filters.

    - SOTF: -255 to 255 (for quicker test: 0-64)
    - SOFM: -127 to 127 (for quicker test: 0-63 or 31-63)
    - SCF: -127 to 127 (for quicker test: 0-63)

13. Next, optimize Close loss/profit (e.g. 0-300 with step 20), then Close time (e.g. 0 to -50 with step -10). These params can be combined with genetic algorithm.
14. Tasks should be optimized at the end. Equity based conditions are not recommended when account is used by multiple EAs. You can optimize by combining conditions and actions together.
15. Next, risk management from 1 to 10.
16. Final touch can be re-optimizing/improving individual strategies or filters per run  (one by one, narrowing down to smaller selected ranges).
17. Re-test using longer periods (2019-2021). Earlier years doesn't need to be profitable.

To summarize, the general order of optimization is to reset filters to defaults and set tasks to none. First optimize strategies separately per timeframe with their best stops (for the same timeframe). Next with all selected strategies try to improve filters, then revise strategies again (after filtering changes), then optionally add tasks at the end (which works better). Compare final results with different modelling and re-run the final setup (broker or platform) with longer periods.

### Rider

Optimization from scratch (recommended period: 2019-2021):

1. Similar process to Advanced, reset all filters to their optimal defaults (don't over complicate filters at the first step, they need to stay simple to not influence strategies main logic). Set SOFM=32 (recommended for grid-like trading).
2. Set 1st task: Eq.>1% then Close orders in profit. 2nd task it's up to you (you can disable or set Eq. <%5, close in not trend). This is our main logic to close orders only when we're in profit. This logic is different comparing to Advanced version, as Rider aims to keep orders for longer time (days or weeks) till certain condition is matched (that's the idea behind this version).
3. Set some default, average optimal stop as a starting point (e.g. MA on H4 or H3). This will apply to all strategies. MA is a good approach to start, to not influence trading strategies by some specific indicator rules (in this case the average of the price is a good approach). 
4. For RIder, it's recommended to set SOSF=2 (strategy filter - open minute-based trades based on hourly-based signals).
4. Optimize strategies per timeframe at a time (disabling all other). E.g. strategies on H1, and disable all other timeframes.
5. Strategies timeframe recommended optimization order: H1, H4, M30, M15, M5, M1.
6. Re-test all strategies how they run together.
7. Optimize strategies by selecting better one (one per timeframe), once other are still running.
8. Re-optimize strategy stops and its timeframe (can be optimized together). Timeframe range (for quicker run): H1 to D1.
8. Re-optimize tasks.
9. Re-optimize filters (one by one, narrowing down to smaller selected ranges).
10. Compare results with different tick modellings (OHLC vs Every ticks) and between different platforms (if required).
11. Re-test using longer periods (2018-2021). Earlier years doesn't need to be profitable, it's just to check how EA deals with risk management for unknown periods (to which was not optimized for).

To summarize, the general order of optimization is to reset filters to defaults, and set tasks to desired outcome (e.g. Equity>1%). First optimize strategies separately per timeframe (using a startup higher-timeframe stop), then with all selected strategies find their best stop (H1-D1). Next try to improve filters, then revise strategies again (after filtering changes). Finally optimize tasks at the end (which works better). Compare final results with different modelling and re-run the final setup (broker or platform) with longer periods.

**Reply by Harry040** [reply]
Advance default parameters:
  float order_close_loss;          // Order close loss (in pips).
  float order_close_profit;        // Order close profit (in pips).
For above two parameters, Why use EA_OrderCloseLoss=40, EA_OrderCloseProfit=0 override per startegy? Just use strategy close signal to exit?

I don't see the "price_profit_method" detail logic. There is nothing about optimizing take profit. 
How to optimize take profit for per strategy? @kenorb

**Reply by kenorb** [reply]
> Why use EA_OrderCloseLoss=40, EA_OrderCloseProfit=0 override per startegy? Just use strategy close signal to exit?

Each strategy has OrderCloseLoss=40 and OrderCloseTime=-30 by default to prevent overfitting in order to avoid too high profits and opening prolonged orders. This is overridden by the main EA params.
The aim is to focus on the opening signal as a base line based on the available technical indicator data. Also if we won't close orders quick enough by fixed loss=40 or after 30 candles, optimization process will fit into specific big market moves which were previously crashes (for example, check out the results for [EA31337-Libre v1.007](https://github.com/EA31337/EA31337-Libre/discussions/25#discussioncomment-1392625) where a [Covid-crash](https://en.wikipedia.org/wiki/2020_stock_market_crash) became a new driving force for large (fake) profits leaving everything else behind).
So it's good to create a harsh and consistent environment which can be spread across the whole period when doing optimization. The more orders are opened the better (e.g. what gives more confidence, a single 5-day long, or 50 trades per week?). Once indicator opening signal params are optimized, then closing orders earlier is no longer necessary.

> I don't see the "price_profit_method" detail logic.

Doesn't seems to be implemented yet. I'm going to check what it suppose to be.

> How to optimize take profit for per strategy?

You can optimize take profit for all strategies, per strategy (in Advanced) you can select different strategy stops (TP/SL) for each strategy.

**Reply by Harry040** [reply]
Appreciate for the comprehensive replay!

**Comment by kenorb** [comment]
For the latest version (v2.011 or similar), the optimization method is quiet simple (assuming you're going to ignore filters).
I usually optimize in MT5 (which support multi-core and the latest ticks) by selecting each strategy one by one starting with all disabled (which is much quicker rather than optimizing multiple params at the same time).
So I start from H1 (on empty set with all strategies disabled). You can also start from H8 towards lower timeframes as well.

![photo_2022-12-14_20-53-39](https://user-images.githubusercontent.com/266306/207712354-14560969-5f2f-41fe-a834-c5fede10a52f.jpg)

Once best strategy is selected, I'm optimizing the next one on H2 then adding it on top, the add another one from H3, till I've a full set.

![photo_2022-12-14_20-53-42](https://user-images.githubusercontent.com/266306/207712421-49b2cebd-a50f-40d2-84df-07369842a57c.jpg)


At the end I can add lower timeframes on top of it to see if they make it better or worse.

In the screen below, you can see Rider v2.011 optimization results (by optimizing and adding H8 strategy on top of the other, which takes around an hour on a good multi-core machine for 2017-2022).

![photo_2022-12-14_20-53-44](https://user-images.githubusercontent.com/266306/207712493-8ad0619f-87ad-42e0-b50c-b426eb6d6036.jpg)

For quicker optimization process, use 1 minute OHLC modelling.

![photo_2022-12-14_20-53-40](https://user-images.githubusercontent.com/266306/207712928-7393606e-e770-457a-adee-d558d08caff7.jpg)

Few more hints:

- The more strategies you add, slightly increase deposit, to avoid any bottlenecks in case some strategies won't be able to prove themselves, because used margin was too high so new orders were prevented.
- After each test, consider changing a starting date (1 month before or after), this will increase randomness by avoiding repeating certain patterns for each test which happened in the past (such as lucky wins). You cannot count on such luck in the real market.
- For Advanced, the same method use when optimizing stop strategies.
